Mathlib/LinearAlgebra/Span.lean

```diff
@@ -0,0 +1,1075 @@
+
+section
+
+variable (R) (M) [Semiring R] [AddCommMonoid M] [Module R M]
+
+/-- Given an element `x` of a module `M` over `R`, the natural map from
+    `R` to scalar multiples of `x`.-/
+@[simps!]
+def toSpanSingleton (x : M) : R →ₗ[R] M :=
+  LinearMap.id.smulRight x
+#align linear_map.to_span_singleton LinearMap.toSpanSingleton
+
+/-- The range of `toSpanSingleton x` is the span of `x`.-/
+theorem span_singleton_eq_range (x : M) : (R ∙ x) = range (toSpanSingleton R M x) :=
+  Submodule.ext fun y => by
+    refine' Iff.trans _ LinearMap.mem_range.symm
+    exact mem_span_singleton
+#align linear_map.span_singleton_eq_range LinearMap.span_singleton_eq_range
+
+-- @[simp] -- Porting note: simp can prove this
+theorem toSpanSingleton_one (x : M) : toSpanSingleton R M x 1 = x :=
+  one_smul _ _
+#align linear_map.to_span_singleton_one LinearMap.toSpanSingleton_one
+
+@[simp]
+theorem toSpanSingleton_zero : toSpanSingleton R M 0 = 0 := by
+  ext
+  simp
+#align linear_map.to_span_singleton_zero LinearMap.toSpanSingleton_zero
+
+end
+
+section AddCommMonoid
+
+variable [Semiring R] [AddCommMonoid M] [Module R M]
+
+variable [Semiring R₂] [AddCommMonoid M₂] [Module R₂ M₂]
+
+variable {σ₁₂ : R →+* R₂}
+
+/-- If two linear maps are equal on a set `s`, then they are equal on `submodule.span s`.
+
+See also `linear_map.eq_on_span'` for a version using `set.eq_on`. -/
+theorem eqOn_span {s : Set M} {f g : M →ₛₗ[σ₁₂] M₂} (H : Set.EqOn f g s) ⦃x⦄ (h : x ∈ span R s) :
+    f x = g x := by refine' span_induction h H _ _ _ <;> simp (config := { contextual := true })
+#align linear_map.eq_on_span LinearMap.eqOn_span
+
+/-- If two linear maps are equal on a set `s`, then they are equal on `submodule.span s`.
+
+This version uses `set.eq_on`, and the hidden argument will expand to `h : x ∈ (span R s : set M)`.
+See `linear_map.eq_on_span` for a version that takes `h : x ∈ span R s` as an argument. -/
+theorem eqOn_span' {s : Set M} {f g : M →ₛₗ[σ₁₂] M₂} (H : Set.EqOn f g s) :
+    Set.EqOn f g (span R s : Set M) :=
+  eqOn_span H
+#align linear_map.eq_on_span' LinearMap.eqOn_span'
+
+/-- If `s` generates the whole module and linear maps `f`, `g` are equal on `s`, then they are
+equal. -/
+theorem ext_on {s : Set M} {f g : M →ₛₗ[σ₁₂] M₂} (hv : span R s = ⊤) (h : Set.EqOn f g s) : f = g :=
+  LinearMap.ext fun _ => eqOn_span h (eq_top_iff'.1 hv _)
+#align linear_map.ext_on LinearMap.ext_on
+
+/-- If the range of `v : ι → M` generates the whole module and linear maps `f`, `g` are equal at
+each `v i`, then they are equal. -/
+theorem ext_on_range {ι : Type _} {v : ι → M} {f g : M →ₛₗ[σ₁₂] M₂} (hv : span R (Set.range v) = ⊤)
+    (h : ∀ i, f (v i) = g (v i)) : f = g :=
+  ext_on hv (Set.forall_range_iff.2 h)
+#align linear_map.ext_on_range LinearMap.ext_on_range
+
+end AddCommMonoid
+
+section Field
+
+variable [Field K] [AddCommGroup V] [Module K V]
+
+open Classical
+
+-- Porting note: TODO Erase this line. Needed because we don't have η for classes.
```

Comment:
```suggestion
-- Porting note: TODO Erase this line. Needed because we don't have η for classes (lean4#2074).
```
* https://github.com/leanprover/lean4/issues/2074